syntax = "proto3";

package protocol;

message Host {
    string name = 1;
    string inner_ip = 2;
    string outer_ip = 3;
    string deploy_ip = 4;
}

message HostCfg {
    repeated Host hosts = 1;
}

message ArtifactFile {
    string src = 1;
    string dst = 2;
    bool is_dir = 3;
    bool need_backup = 4;
    repeated string tags = 5;
}

message AgentArtifact {
    repeated ArtifactFile files = 1;
    string name = 2;
    int32 function_id = 3;
    string start_cmd = 4;
    string stop_cmd = 5;
    string clean_cmd = 6;
    string check_cmd = 7;
}

message TemplateArtifactItem {
    string template_name = 1;
    repeated ArtifactFile files = 2;
}

message TemplateArtifact {
    repeated TemplateArtifactItem template_artifacts = 1;
}

message NormalArtifactItem {
    string template_name = 1;
    string name = 2;
    int32 function_id = 3;
    int32 start_priority = 4;
    repeated ArtifactFile files = 5;
    uint32 copies = 6;
}

message NormalArtifact {
    repeated NormalArtifactItem artifacts = 1;
}

message ArtifactGroupItem {
    string name = 1;
    repeated string artifact_names = 2;
}

message ArtifactGroup {
    repeated ArtifactGroupItem artifact_groups = 1;
}

message KeyValue {
    string key = 1;
    string value = 2;
}

message InstanceGroup {
    string artifact_group_name = 1;
    string host_name = 2;
    int32 instance_id = 3;
}

message Zone {
    string name = 1;
    int32 id = 2;
    repeated InstanceGroup instance_groups = 4;
}

message World {
    string name = 1;
    int32 id = 2;
    string user = 3;
     string passwd = 4;
    repeated KeyValue variables = 5;
    repeated Zone zones = 6;
}

message DeployInfo {
    string tmp_root_path = 2;
    string dst_root_path = 3;
    int32 listen_port = 4;
}

message Deploy {
    string tmp_root_path = 2;
    string dst_root_path = 3;
    repeated KeyValue variables = 4;
    repeated World worlds = 5;
    int32 listen_port = 6;
}

